### PR TITLE
Add the Google Plugin!

### PR DESCRIPTION
**Pull request by gourabnagDev**

### CODE REVIEW

File: src/Chat/Plugin/Chuck.php [hunk 1]
```diff
@@ -36,8 +36,6 @@ private function getResult(Message $message): \Generator {
 
         $result = json_decode($response->getBody(), true);
 
-        $this->chatClient->postMessage("not works :(");
```
Comment (staabm) [inline src/Chat/Plugin/Chuck.php]: Unrelated change?

File: src/Chat/Plugin/Google.php [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                "It was google's fault, not mine."
+            );
+
+            return;
+        }
+
+        $internalErrors = libxml_use_internal_errors(true);
+        $dom = new \DOMDocument();
+        $google = $response->getBody();
+        mb_convert_encoding($google, 'utf-8', mb_detect_encoding($google));
```
Comment (staabm) [inline src/Chat/Plugin/Google.php]: Google also uses utf8, dont they?

File: src/Chat/Plugin/Google.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            );
+        }
+        $to_post_message .= "  **[Google Search Url]($uri)**";
+
+        $to_post_message = str_replace("\r", " ", $to_post_message);
+        $to_post_message = str_replace("\r\n", " ", $to_post_message);
+        try {
+            yield from $this->chatClient->postMessage(str_replace("\n", " ", $to_post_message));
+        } catch(Exception $ex) {
+            yield from $this->chatClient->postMessage("*cough, cough* uhm, we faced a problem here... Could you take a look?");
```
Comment (staabm) [inline src/Chat/Plugin/Google.php]: Should you log the exception?

Comment (staabm) [inline src/Chat/Plugin/Google.php]: Actually.. i guess this postMessage in the catch block will not work when a above postMessage didnt work?

Comment (gourabnagDev) [inline src/Chat/Plugin/Google.php]: didn't know that...

File: src/Chat/Plugin/Google.php (hunk 2 above)
Comment (gourabnagDev) [inline src/Chat/Plugin/Google.php]: do they? I dunno tho, should I remove that line?

File: src/Chat/Plugin/Chuck.php (hunk 1 above)
Comment (gourabnagDev) [inline src/Chat/Plugin/Chuck.php]: yes

File: src/Chat/Plugin/Google.php (hunk 3 above)
Comment (staabm) [inline src/Chat/Plugin/Google.php]: Why did you add that catch block in the first place?

File: src/Chat/Plugin/Google.php (hunk 2 above)
Comment (staabm) [inline src/Chat/Plugin/Google.php]: What does mb_detect_encoding return here for a google result?

File: src/Chat/Plugin/Google.php [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    private function getResult(Message $message): \Generator {
+        $uri = "https://www.google.com/search?q=" . urlencode(implode(' ', $message->getParameters()));
+
+        /** @var Response $response */
+        $response = yield from $this->chatClient->request($uri);
+
+        if ($response->getStatus() !== 200) {
+            yield from $this->chatClient->postMessage(
+                "It was google's fault, not mine."
```
Comment (kelunik) [inline src/Chat/Plugin/Google.php]: `google` → `Google`

File: src/Chat/Plugin/Google.php [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                ":%s Did you know? That `%s...` dosen't exist in the world! Cuz' GOOGLE can't find it :P",
+                $message->getOrigin(),
+                substr(implode(" ", $message->getParameters()), 0, 60)
+            ));
+            return;
+        }
+        $length = 3;
+        if($nodes->length < $length) {
+            $length = $nodes->length;
+        }
```
Comment (kelunik) [inline src/Chat/Plugin/Google.php]: Maybe just simplify to `$length = min(3, $nodes->length)`

File: src/Chat/Plugin/Google.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+        $to_post_message = "";
```
Comment (kelunik) [inline src/Chat/Plugin/Google.php]: Should use `camelCase`, so `$toPostMessage`.

File: src/Chat/Plugin/Google.php [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        for($i = 0; $i < $length; $i++) {
+            start_of_loop:
+            $current_node = $nodes[$i];
+            $node_link = $xpath->query("//h3/a", $current_node);
+            $node_link_text = $node_link->item($i)->textContent;
+            $node_link = $node_link->item($i)->getAttribute("href");
+            $node_description = substr(strip_tags(nl2br($xpath->query('//span[@class="st"]', $current_node)->item($i)->textContent)), 0, 55);
+            if(preg_match('~^/url\?q=([^&]*)~', $node_link, $matches) == false) {
+                ++$i;
+                goto start_of_loop;
```
Comment (kelunik) [inline src/Chat/Plugin/Google.php]: Replace with `continue;`.

File: src/Chat/Plugin/Google.php [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+            }
+            $link = $matches[1];
+            $api_uri = sprintf(
+                "https://api-ssl.bitly.com/v3/shorten?access_token=%s&longUrl=%s",
+                "5c8c24601d7c44563e56378dc81300cfd27f0cd3",
```
Comment (kelunik) [inline src/Chat/Plugin/Google.php]: Uhm, public access token? Should probably be something in the config then.

File: src/Chat/Plugin/Google.php [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+                $link
+            );
+            $shortener = yield from $this->chatClient->request($api_uri);
+            $shortened = json_decode($shortener->getBody(), true);
+            $shortenedLink = $shortened["data"]["url"];
+            $to_post_message .= sprintf(
+                "  **[%s](%s)** %s...  |",
+                utf8_encode($node_link_text),
```
Comment (kelunik) [inline src/Chat/Plugin/Google.php]: I don't think we need that call to `utf8_encode`?

File: src/Chat/Plugin/Google.php (hunk 3 above)
Comment (kelunik) [inline src/Chat/Plugin/Google.php]: Catch block should be removed, yes.

File: src/Chat/Plugin/Google.php (hunk 7 above)
Comment (kelunik) [inline src/Chat/Plugin/Google.php]: You can also remove the `++$i` then.

File: src/Chat/Plugin/Google.php (hunk 9 above)
Comment (gourabnagDev) [inline src/Chat/Plugin/Google.php]: Otherwise, some chars won't be rendered properly if I removed them

**Comment by gourabnagDev** [comment]
Would, fix all those :+1: but atm I gtg...

File: src/Chat/Plugin/Google.php (hunk 4 above)
Comment (staabm) [inline src/Chat/Plugin/Google.php]: Google's -> Googles'

File: src/Chat/Plugin/Google.php (hunk 9 above)
Comment (kelunik) [inline src/Chat/Plugin/Google.php]: In that case you should make sure to use the right encoding then passing the HTML to `DOMDocument`, not when pulling texts out.

File: src/Chat/Plugin/Google.php (hunk 4 above)
Comment (kelunik) [inline src/Chat/Plugin/Google.php]: @staabm Google's is right. Google is singular.

**Comment by gourabnagDev** [comment]
Done, all of em'!

@kelunik @staabm

File: src/Chat/Plugin/Chuck.php (hunk 1 above)
Comment (kelunik) [inline src/Chat/Plugin/Chuck.php]: This is still there and has to be removed.

File: src/Chat/Plugin/Google.php [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+                "https://api-ssl.bitly.com/v3/shorten?access_token=%s&longUrl=%s",
+                $this->bitlyAccessToken,
+                $link
+            );
+            $shortener = yield from $this->chatClient->request($apiUri);
+            $shortened = json_decode($shortener->getBody(), true);
+            $shortenedLink = $shortened["data"]["url"];
+            $toPostMessage .= sprintf(
+                "  **[%s](%s)** %s...  |",
+                utf8_encode($nodeLinkText),
```
Comment (kelunik) [inline src/Chat/Plugin/Google.php]: If `$nodeLinkText` needs `utf8_encode`, why doesn't `$nodeDescription` not need it? Did you try to apply `utf8_encode` in line 40, where you call `loadHTML`?

File: src/Chat/Plugin/Google.php [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+                ":%s Did you know? That `%s...` dosen't exist in the world! Cuz' GOOGLE can't find it :P",
+                $message->getOrigin(),
+                substr(implode(" ", $message->getParameters()), 0, 60)
+            ));
+            return;
+        }
+        $length = min(3, $nodes->length);
+        $toPostMessage = "";
+        for($i = 0; $i < $length; $i++) {
+            start_of_loop:
```
Comment (kelunik) [inline src/Chat/Plugin/Google.php]: Label should be removed, since the `goto` is gone now.

Comment (gourabnagDev) [inline src/Chat/Plugin/Google.php]: Oh well my bad :( didn't notice those... Gonna fix them in a while..

Gourab Nag

File: src/Chat/Plugin/Google.php [hunk 12]
```diff
@@ -49,37 +59,59 @@ private function getResult(Message $message): \Generator {
             ));
             return;
         }
+
         $length = min(3, $nodes->length);
         $toPostMessage = "";
+
         for($i = 0; $i < $length; $i++) {
-            start_of_loop:
+            var_dump($toPostMessage);
```
Comment (PeeHaa) [inline src/Chat/Plugin/Google.php]: Left over dump

File: src/Chat/Plugin/Google.php [hunk 13]
```diff
@@ -49,37 +59,59 @@ private function getResult(Message $message): \Generator {
...
+        $googleSearchBitlyLink = json_decode($googleSearchBitlyLink->getBody(), true)["data"]["url"];
+        if(strlen($toPostMessage) + strlen("  **[Search Url]($googleSearchBitlyLink)**") > 500) {
+            $toPostMessage = substr($toPostMessage, 0, 500 - (strlen("  **[Search Url]($googleSearchBitlyLink)**") + 1));
+            $toPostMessage .= "|";
         }
-        $toPostMessage .= "  **[Google Search Url]($uri)**";
+        $toPostMessage .= "  **[Search Url]($googleSearchBitlyLink)**";
         $toPostMessage = str_replace("\r", " ", $toPostMessage);
         $toPostMessage = str_replace("\r\n", " ", $toPostMessage);
+        var_dump($toPostMessage);
```
Comment (PeeHaa) [inline src/Chat/Plugin/Google.php]: Left over dump

File: src/Chat/Plugin/Google.php [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+                substr(implode(" ", $message->getParameters()), 0, 60)
+            ));
+            return;
+        }
+
+        $length = min(3, $nodes->length);
+        $toPostMessage = "";
+
+        for($i = 0; $i < $length; $i++) {
+            var_dump($toPostMessage);
```
Comment (oldskool) [inline src/Chat/Plugin/Google.php]: Seems like some debug stuff?

File: src/Chat/Plugin/Google.php [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+        $googleSearchBitlyLink = yield from $this->chatClient->request($apiUri);
+        $googleSearchBitlyLink = json_decode($googleSearchBitlyLink->getBody(), true)["data"]["url"];
+        if(strlen($toPostMessage) + strlen("  **[Search Url]($googleSearchBitlyLink)**") > 500) {
+            $toPostMessage = substr($toPostMessage, 0, 500 - (strlen("  **[Search Url]($googleSearchBitlyLink)**") + 1));
+            $toPostMessage .= "|";
+        }
+        $toPostMessage .= "  **[Search Url]($googleSearchBitlyLink)**";
+        $toPostMessage = str_replace("\r", " ", $toPostMessage);
+        $toPostMessage = str_replace("\r\n", " ", $toPostMessage);
+        var_dump($toPostMessage);
```
Comment (oldskool) [inline src/Chat/Plugin/Google.php]: And more debug stuff?

File: src/Chat/Plugin/Google.php [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+        $google = utf8_encode($response->getBody());
+        $dom->loadHTML($google);
+        libxml_use_internal_errors($internalErrors);
+        $xpath = new \DOMXPath($dom);
+
+        $nodes = $xpath->query("//*[contains(concat(' ', normalize-space(@class), ' '), ' g ')]");
+
+        if($nodes->length === 0) {
+            yield from $this->chatClient->postMessage(sprintf(
+                ":%s Did you know? That `%s...` dosen't exist in the world! Cuz' GOOGLE can't find it :P",
```
Comment (oldskool) [inline src/Chat/Plugin/Google.php]: Typo: **dosen't** instead of *doesn't*.

**Comment by gourabnagDev** [comment]
@PeeHaa @oldskool ^

**Comment by oldskool** [comment]
It apparently still needs a rebase, but other than that it looks good to me.
